### PR TITLE
Document new utf8 option of Routing component

### PR DESCRIPTION
**Pull request by mickaelandrieu**
Hi,

all credits to @javiereguiluz and @nicolas-grekas, I've just copy/pasted the blog article http://symfony.com/blog/new-in-symfony-3-2-unicode-routing-support#comment-form

Mickaël

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
See #6890
we now have to decide which one is better :)

**Comment by nicolas-grekas** [comment]
or merge both since they aren't on the same page

**Comment by mickaelandrieu** [comment]
@nicolas-grekas my bad, didnt thank you have started something... as you like, I wont be sad if you close this one my only interest is to see this feature documented.

By the way, big thanks! From a business/e-commerce point of view, this feature is so useful :)

File: components/routing.rst [hunk 1]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
     are saved in the ``cache_dir``. This means your script must have write
     permissions for that location.
 
+Unicode routing support
```
Comment (xabbuh) [inline components/routing.rst]: Unicode Routing Support

File: components/routing.rst [hunk 2]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The routing component supports UTF-8 characters in route paths and requirements. Thanks to the `utf8` route option, you can make Symfony match and generate routes with UTF-8 characters:
```
Comment (xabbuh) [inline components/routing.rst]: Routing component

Comment (xabbuh) [inline components/routing.rst]: The option needs to be enclosed by double backticks.

File: components/routing.rst [hunk 3]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+.. code-block:: yaml
+
+    # routes.yml
+    route1:
+        path:     /category/{name}
+        defaults: { _controller: 'DefaultController::categoryAction' }
+        requirements:
+          name: ".+"
+        options:
+          utf8: true
```
Comment (xabbuh) [inline components/routing.rst]: Please add Annotations, XML, and PHP examples too.

File: components/routing.rst [hunk 4]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+
+In this route, the `utf8` option set to true makes Symfony consider the `.` requirement to match any UTF-8 characters instead of just a single byte character, so the following URLs would match: `/category/日本語`, `/category/فارسی`, `/category/한국어`, etc. In case you are wondering, this option also allows to include and match emojis in URLs.
```
Comment (xabbuh) [inline components/routing.rst]: please use double backticks everywhere here too

Comment (xabbuh) [inline components/routing.rst]: and please wrap lines after the first word that crosses the 72nd character

File: components/routing.rst (hunk 2 above)
Comment (xabbuh) [inline components/routing.rst]: Please also wrap lines after the first word that crosses the 72nd character.

File: components/routing.rst [hunk 5]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+.. code-block:: yaml
+
+    # routes.yml
+    route2:
+        path:     /default/{default}
+        defaults: { _controller: 'DefaultController::defaultAction' }
+        requirements:
+          default: "日本語|فارسی"
+        options:
+          utf8: true
```
Comment (xabbuh) [inline components/routing.rst]: We need other format examples here too.

File: components/routing.rst [hunk 6]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+
+This second example describes how you can use utf8 string as a routing requirement.
```
Comment (xabbuh) [inline components/routing.rst]: UTF-8

File: components/routing.rst [hunk 7]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+
+.. note::
+
+    In Symfony 3.2 there is no need to set this utf8 explicitly. As soon as Symfony finds a UTF-8 character in the route path or requirements, it will turn the UTF-8 support automatically. In addition to UTF-8 characters, the Routing component also supports all the `PCRE Unicode properties`_, which are escape sequences that match generic character types. For example, `\p{Lu}` matches any uppercase character in any language, `\p{Greek}` matches any Greek character, `\P{Han}` matches any character not included in the Chinese Han script.
```
Comment (xabbuh) [inline components/routing.rst]: same as above

File: components/routing.rst [hunk 8]
```diff
@@ -346,6 +346,44 @@ automatically in the background if you want to use it. A basic example of the
...
+  
+.. versionadded:: 3.2
+    The UTF8 support in route paths and requirements were added in Symfony 3.2
```
Comment (xabbuh) [inline components/routing.rst]: UTF-8 support for route paths and requirements were introduced in Symfony 3.2.

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by mickaelandrieu** [comment]
Wow ... I feel so bad :(  thank you for this review, I will address comments tomorrow!

**Comment by xabbuh** [comment]
No need to worry. There's not much to change. :)

File: components/routing.rst (hunk 5 above)
Comment (mickaelandrieu) [inline components/routing.rst]: Do we really need ? I don't see the gain here when it's already well described [here](http://symfony.com/doc/current/routing.html#routing-examples)

I will do it if you insist, though (sounds like a simple copy/paste of the resource I've linked ^ ).

**Comment by mickaelandrieu** [comment]
Hi @xabbuh,

I've adressed almost all comments except the one about complete all formats, let me know if this need more work :)

Mickaël

File: components/routing.rst (hunk 5 above)
Comment (nicolas-grekas) [inline components/routing.rst]: That's how the doc is yes, config are always in all formats, repeating is teaching isn't it?

Comment (mickaelandrieu) [inline components/routing.rst]: true, let's go for it :)

**Comment by xabbuh** [comment]
👍 (except for the missing config formats)

File: components/routing.rst [hunk 9]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+        <!-- app/config/routing.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <routes xmlns="http://symfony.com/schema/routing"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/routing
+                http://symfony.com/schema/routing/routing-1.0.xsd">
+
+            <route id="route1" path="/category/{name}">
+                <default key="_controller">AppBundle:Default:category</default>
+                <option key="utf8">true</option>
```
Comment (mickaelandrieu) [inline components/routing.rst]: Question here: how can we do to ensure true will be converted to boolean value ?

Comment (xabbuh) [inline components/routing.rst]: That's the responsibility of the loader.

Comment (mickaelandrieu) [inline components/routing.rst]: So ... is it good, or does it need much work from me ?

File: components/routing.rst [hunk 10]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
     are saved in the ``cache_dir``. This means your script must have write
     permissions for that location.
 
+Unicode Routing support
```
Comment (xabbuh) [inline components/routing.rst]: Support

File: components/routing.rst [hunk 11]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+            }
+
+  .. code-block:: yaml
+
+      # routes.yml
+      route1:
+          path:     /category/{name}
+          defaults: { _controller: 'DefaultController::categoryAction' }
+          options:
+            utf8: true
```
Comment (xabbuh) [inline components/routing.rst]: please indent this line with four spaces

File: components/routing.rst [hunk 12]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+            {
+                // ...
+            }
+
+  .. code-block:: yaml
+
+      # routes.yml
+      route1:
+          path:     /category/{name}
+          defaults: { _controller: 'DefaultController::categoryAction' }
```
Comment (xabbuh) [inline components/routing.rst]: `AppBundle:Default:category`

File: components/routing.rst (hunk 9 above)
Comment (xabbuh) [inline components/routing.rst]: It's good as it is. :)

File: components/routing.rst [hunk 13]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+      array(
+          'utf8' => true,
+      ));
+
+      // ...
+
+      return $collection;
+
+
+In this route, the ``utf8`` option set to true makes Symfony consider the
```
Comment (xabbuh) [inline components/routing.rst]: please enclose `true` in double backticks

File: components/routing.rst [hunk 14]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+``.`` requirement to match any UTF-8 characters instead of just a single
+byte character, so the following URLs would match: ``/category/日本語``,
```
Comment (xabbuh) [inline components/routing.rst]: maybe split this here: "[...] character. This means that the following URL would match [...]"

File: components/routing.rst [hunk 15]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+``/category/فارسی``, ``/category/한국어``, etc. In case you are wondering,
+this option also allows to include and match emojis in URLs.
+
+.. configuration-block::
+
+  .. code-block:: php-annotations
```
Comment (xabbuh) [inline components/routing.rst]: please indent with four spaces

File: components/routing.rst [hunk 16]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+      array(
+          'utf8' => true,
+      ));
+
+      // ...
+
+      return $collection;
+
+
+This second example describes how you can use UTF-8 string as a routing
```
Comment (xabbuh) [inline components/routing.rst]: strings

Comment (xabbuh) [inline components/routing.rst]: [...] as routing requirements.

File: components/routing.rst [hunk 17]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+requirement.
+
+.. note::
+
+    In Symfony 3.2 there is no need to set this utf8 explicitly. As soon
```
Comment (xabbuh) [inline components/routing.rst]: [...] to explicitly set the `utf8` option.

File: components/routing.rst [hunk 18]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+    it will turn the UTF-8 support automatically. In addition to UTF-8
+    characters, the Routing component also supports all the `PCRE Unicode properties`_,
+    which are escape sequences that match generic character types. For
+    example, ``\p{Lu}`` matches any uppercase character in any language,
+    ``\p{Greek}`` matches any Greek character, ``\P{Han}`` matches any character
+    not included in the Chinese Han script.
+
+.. versionadded:: 3.2
+    UTF-8 support for route paths and requirements were introduced in
+    Symfony 3.2.
```
Comment (xabbuh) [inline components/routing.rst]: I would move this up to the beginning of this section.

**Comment by mickaelandrieu** [comment]
@xabbuh thanks again for your time. Comments addressed ;)

File: components/routing.rst [hunk 19]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
     are saved in the ``cache_dir``. This means your script must have write
     permissions for that location.
 
+Unicode Routing Support
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 3.2
+    UTF-8 support for route paths and requirements were introduced in
```
Comment (xabbuh) [inline components/routing.rst]: was

File: components/routing.rst [hunk 20]
```diff
@@ -346,6 +346,181 @@ automatically in the background if you want to use it. A basic example of the
...
+            }
+
+  .. code-block:: yaml
+
+      # routes.yml
+      route1:
+          path:     /category/{name}
+          defaults: { _controller: 'AppBundle:Default:category' }
+          options:
+            utf8: true
```
Comment (xabbuh) [inline components/routing.rst]: indentation :)

Comment (xabbuh) [inline components/routing.rst]: Looks like it's the same in the other YAML example too.

**Comment by xabbuh** [comment]
@mickaelandrieu I left two last minor comments and indentation of the PHP code examples looks weird too. But overall this change looks ready to be merged. 👍

**Comment by mickaelandrieu** [comment]
@xabbuh I won't give up now :D let me improve again my contribution ;)

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
👍 

Status: Reviewed

**Comment by wouterj** [comment]
Thanks @mickaelandrieu! After your quick responses to the reviews, this PR has now made it to the docs!

Fyi, I've fixed some indentation issues and splitted the big note directive in c5328d7.

**Comment by mickaelandrieu** [comment]
@wouterj  thank you, it's better now :)
Always happy to help ;)

**Pull request merged**